### DISCUSSION TITLE
Two actions, single state update

### DISCUSSION THREAD

**Discussion by PavelZhuravlev** [pmndrs/zustand#1583] [Q&A]
Hey!
I'm developing a medium sized project with zustand and it's based on slices pattern described here - https://github.com/pmndrs/zustand/blob/main/docs/guides/slices-pattern.md

I have a question about the technique of calling a sub-action to break a more complex action into parts:
```
export const createBearFishSlice = (set) => ({
  addBearAndFish: () => {
    createBearSlice(set).addBear()
    createFishSlice(set).addFish()
  },
})
```

I really love this approach because in the real-world example there are a lot of API methods that change the state in different slices, often these state changes are repeated through multiple actions, BUT

Is there any way to merge these sub-actions into a single update in the state, instead of executing them one by one?

Current behaviour:

<img width="542" alt="Снимок экрана 2023-01-28 в 21 18 16" src="https://user-images.githubusercontent.com/14184094/215286902-b1df048b-6cda-4ab9-8178-db85d2cc7557.png">
<img width="544" alt="Снимок экрана 2023-01-28 в 21 20 03" src="https://user-images.githubusercontent.com/14184094/215286906-3cacf4ad-e737-4e8f-9466-c0b6c8b37871.png">

Expected behaviour:
<img width="583" alt="Снимок экрана 2023-01-28 в 21 21 42" src="https://user-images.githubusercontent.com/14184094/215286951-dc85faed-03fe-4557-a3ef-cc413d3c15d3.png">

of course i can achieve it by updating state directly using `set`, but i'm curious can I do the same by batching/merging multiple actions together?

often there is a need to call the same state updates as a result of different side effects, and mutation can be quite large (especially without immer), so you want to be able to encapsulate these mutations in sub actions and call them in higher level actions without having an intermediate state of the application and not breaking time travel

If needed, I can provide real world example.
Thank you in advance for your answer, let me know if I'm thinking in the wrong direction.

**Comment by PavelZhuravlev** [comment]
demo: https://codesandbox.io/s/magical-sky-ortp5q?file=/src/App.js

**Comment by dai-shi** [comment]
Assuming this is only about the `devtools` middleware, batching multiple `set`s into one action seems technically possible. One concern is we can't explicitly name the action name. Another concern is it will be the entire re-implementation, and keeping all the current features might be hard.

**Comment by PavelZhuravlev** [comment]
@dai-shi thanks for your answer
tbh, I don't really care how it will show up in devtools
I'm looking for the opportunity to batch multiple actions together to guarantee that my UI won't respond to changes in the state until both actions are executed, more like write multiple state updates in the one shot.

similar concept is shown here: https://github.com/tshelburne/redux-batched-actions

**Reply by dai-shi** [reply]
That's totally different story.
Calling `set` multiple times is just fine. It will be batched (by React).
Your UI won't respond to each `set`.
